packages/utils/test/multiaddrToIpNet.spec.ts

```diff
@@ -0,0 +1,18 @@
+/* eslint-env mocha */
+
+import { expect } from 'aegir/chai'
+import { multiaddrToIpNet } from '../src/multiaddrToIpNet.js'
+
+describe('multiaddrToIpNet', () => {
+  it('should convert a simple multiaddr to an IpNet', () => {
+    const ma = '/ip4/127.0.0.1'
+    const ipNet = multiaddrToIpNet(ma)
+    expect(ipNet.toString()).to.equal('127.0.0.1/32')
+  })
+
+  it('should convert a multiaddr with a ipcidr to an IpNet', () => {
+    const ma = '/ip4/127.0.0.1/ipcidr/32'
+    const ipNet = multiaddrToIpNet(ma)
+    expect(ipNet.toString()).to.equal('127.0.0.1/32')
+  })
```

Comment:
Please can you add some tests for IPv6 addresses

Reply:
Done. 
But 
```typescript
it('should convert a simple IPv6 multiaddr to an IpNet', () => {
    const ma = '/ip6/::1/ipcidr/128'
    const ipNet = multiaddrToIpNet(ma)
    expect(ipNet.toString()).to.equal('0000:0000:0000:0000:0000:0000:0000:0001/128')
  })
```
Don't work it gives error so I removed it, **but it should be a basic test it shouldn't fail**

I'm not expert in ip6 but look at this examples: is there a bug in @multiformats/multiaddr ?

```javascript
import { multiaddr} from '@multiformats/multiaddr'

function test (a) {
  const ma = multiaddr(a)
  console.log(a, ma.toString())
}

test('/ip6/0:0:0:0:0:0:0:1') // /ip6/::1
test('/ip6/::1') // /ip6/::1
test('/ip6/2001:db8:3333:4444:5555:6666:7777:8888/ipcidr/99') // /ip6/2001:db8:3333:4444:5555:6666:7777:8888/ipcidr/99
test('/ip6/2001:db8::/ipcidr/99') // /ip6/2001:db8::/ipcidr/99
test('/ip6/2001:db8::/ipcidr/100')// throw new SyntaxError('Unexpected end of data')
test('/ip6/::1/ipcidr/128') // throw new SyntaxError('Unexpected end of data')
test('/ip6/0000:0000:0000:0000:0000:0000:0000:0001/ipcidr/128') // throw new SyntaxError('Unexpected end of data')
test('/ip6/0:0:0:0:0:0:0:1/ipcidr/128') // throw new SyntaxError('Unexpected end of data')
```

```bash
node test.mjs 
/ip6/0:0:0:0:0:0:0:1 /ip6/::1
/ip6/::1 /ip6/::1
/ip6/2001:db8:3333:4444:5555:6666:7777:8888/ipcidr/99 /ip6/2001:db8:3333:4444:5555:6666:7777:8888/ipcidr/99
/ip6/2001:db8::/ipcidr/99 /ip6/2001:db8::/ipcidr/99
file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/multiformats/dist/src/bases/base.js:156
        throw new SyntaxError('Unexpected end of data');
              ^

SyntaxError: Unexpected end of data
    at decode (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/multiformats/dist/src/bases/base.js:156:15)
    at Decoder.decode [as baseDecode] (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/multiformats/dist/src/bases/base.js:199:20)
    at Decoder.decode (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/multiformats/dist/src/bases/base.js:52:25)
    at fromString (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/uint8arrays/dist/src/from-string.node.js:20:25)
    at convertToBytes (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/@multiformats/multiaddr/dist/src/convert.js:100:20)
    at stringToMultiaddrParts (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/@multiformats/multiaddr/dist/src/codec.js:44:23)
    at new Multiaddr (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/@multiformats/multiaddr/dist/src/multiaddr.js:58:21)
    at multiaddr (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/node_modules/@multiformats/multiaddr/dist/src/index.js:195:12)
    at test (file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/test.mjs:4:14)
    at file:///home/luca/Informatica/Learning/PNL_Launchpad_Curriculum/Libp2p/tmp/test-multiformats/test.mjs:12:1

Node.js v22.10.0

```

Reply:
In [@multifomats/multiaddr/src/convert.ts](https://github.com/multiformats/js-multiaddr/blob/d4164fbce730d3b43a10682b651bcc96db845e7c/src/convert.ts)

```typescript
export function convertToBytes (proto: string | number, str: string): Uint8Array {
  const protocol = getProtocol(proto)
  switch (protocol.code) {
    case 4: // ipv4
      return ip2bytes(str)
    case 41: // ipv6
      return ip2bytes(str)
    case 42: // ipv6zone
      return str2bytes(str)

    case 6: // tcp
    case 273: // udp
    case 33: // dccp
    case 132: // sctp
      return port2bytes(parseInt(str, 10))

    case 53: // dns
    case 54: // dns4
    case 55: // dns6
    case 56: // dnsaddr
    case 400: // unix
    case 449: // sni
    case 777: // memory
      return str2bytes(str)

    case 421: // ipfs
      return mh2bytes(str)
    case 444: // onion
      return onion2bytes(str)
    case 445: // onion3
      return onion32bytes(str)
    case 466: // certhash
      return mb2bytes(str)
    case 481: // http-path
      return str2bytes(globalThis.decodeURIComponent(str))
    default:
      return uint8ArrayFromString(str, 'base16') // no clue. convert from hex
  }
}```

There is no case 43 (ipcidr) so in the default case 

The issue arises because the string 100 (or >99) is being interpreted as hexadecimal when passed to uint8ArrayFromString with the base16 encoding. Hexadecimal (base16) encoding expects strings consisting of characters from 0-9 and a-f. In base16, each character represents 4 bits, so the total number of characters must be even to decode into full bytes.

If I add 

```typescript
case 43: // ipcidr
      console.log('convertToBytes case 43:', str)
      str = str.length % 2 !== 0 ? '0' + str : str
      console.log('convertToBytes uint8ArrayFromString', uint8ArrayFromString(str, 'base16'))
      return uint8ArrayFromString(str, 'base16')
```

Reply:
I'm a bit confused by this thread, but this may help:

The CIDR mask is limited to 0-32 for ipv4 and 0-128 for ipv6. When you convert the string form of the CIDR mast to bytes, you only need a single byte to represent all options. So the string to byte logic should convert the given decimal string number to a byte. And vice versa for the bytes to string part. See the go-multiaddr implementation for an example:

https://github.com/multiformats/go-multiaddr/pull/177/files#diff-38293d4bb2018b8660c43e7007e16a293752d0205fdc660558c419d6d3d199a6R59-R72

Reply:
@MarcoPolo 
Make sense. But the existing code is expecting a uint8Array so I was trying to change the less code possibile. Maybe I have to go take a deeper look...

Reply:
@SgtPooki @achingbrain @MarcoPolo 

https://github.com/multiformats/js-multiaddr/pull/394